### ISSUE TITLE
Long path to MEF directories

### ISSUE THREAD

**Issue by nelagri** [msel-source/pymef#21]
Hello Jan,

I have encountered another problem. If the path to MEF file is too long (longer than 256 characters), reading fails. The exception is raised in the 9220th line of meflib.c file. Since MEF files have the directory-like structure, it is possible to ensure that longer paths will be supported?

Thank you.

**Comment by cimbi** [comment]
I'll review this during upcoming days. Should not be a difficult fix.

**Comment by cimbi** [comment]
@nelagri sorry for not addressing this issue earlier. @dancrepeau are you OK with modifications of MEFLIB? I think the only thing that needs to be changed is the number of bytes for MEF_FULL_FILE_NAME_BYTES int meflib.h

**Comment by dancrepeau** [comment]
Sounds okay to me.  That value is currently set to 1024 ascii (or 255 unicode).  Would doubling the value be sufficient?  I would hesitate to go too large, since memory allocations are often done using that value.  We'll also have to be on the look for new bugs that might be introduced.  That shouldn't happen, as long as that same value is used everywhere, but it's possible that in some code somewhere, a value is hard-coded instead of using this defined value.

**Comment by em7** [comment]
I'm afraid of that 2*255 unicode is still 1/64 of what Windows supports. See:

https://docs.microsoft.com/en-us/cpp/c-runtime-library/path-field-limits?view=msvc-160#:~:text=The%20C%20Runtime%20supports%20path,to%20support%20these%20longer%20paths.

Note that "unicode" in the context of Windows API is UTF-16, not UTF-8.

**Comment by dancrepeau** [comment]
Okay - what number would you suggest?  It's set at 1024 bytes now.  I can run some very quick tests with a new number, and if everything looks good, let's make the change.

-Dan

**Comment by em7** [comment]
At least the value of PATH_MAX macro from limits.h header file on a linux-based system, value of _MAX_PATH macro from stdlib.h on Windows and a similar constant on Mac (I don't know MacOS but every operating system has such limit published in its C or C++ API, you need to consult Mac OS API documentation).

Setting such value to an arbitrary number is a very bad programming practice. These constants are a part the OS API for a reason.

**Comment by dancrepeau** [comment]
Understood.  The meflib files are helper utilities, separate from the file format itself.  It was written and tested for a limited number of use-cases on Mac OS, and as we expand to other operating systems and the user base expands, we will need to make changes.  Your information here is very helpful.

**Comment by cimbi** [comment]
@dancrepeau will you take care of this? I can do it but it will take some time before I can get to it.

**Comment by dancrepeau** [comment]
Okay - so my understanding of unicode is that it can be up to 4 bytes per character.  When I use _MAX_PATH in my code, the value is 260.  I think this is pretty standard for Microsoft Windows, at least based on what I'm reading in question forums.  So if we do 4 * _MAX_PATH, that would yield 1040 bytes, which is only a bit bigger than the original 1024.  So I'm wondering if that would help you?  My understanding from the microsoft.com link you mentioned is that the language and the file system support much longer paths, but the operating system does not, at least not in this case.  (I'm using Windows 10).  Let me know if there is something I am missing.  Thanks!

**Comment by em7** [comment]
The Windows operating system supports up to 32,767 characters (not bytes). Filesystems might not. Relative paths are limited to _MAX_PATH on NTFS, absolute paths are a can of worms.

![image](https://user-images.githubusercontent.com/4933100/111067932-cb0dc380-84c6-11eb-8459-d17adf44530e.png)

Both Windows API constant and the Wide-char volume information show the value of 260 or 255. Both values are number of UTF-16 characters. The 255 value is number of characters _between_ slashes. Since many network file systems allow longer names, there is a workaround in the Windows Unicode API, see 

```The Windows API has many functions that also have Unicode versions to permit an extended-length path for a maximum total path length of 32,767 characters. This type of path is composed of components separated by backslashes, each up to the value returned in the lpMaximumComponentLength parameter of the GetVolumeInformation function (this value is commonly 255 characters). To specify an extended-length path, use the "\\?\" prefix. For example, "\\?\D:\very long path".```

https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation#:~:text=The%20Windows%20API%20has%20many,path%20length%20of%2032%2C767%20characters.

You simply can not set this to an arbitrary number, you should always follow the API of the underlying operating system. Either by usage of a 3rd party multi-platfom library or a series of `#ifdef`s . Or, what I would strongly suggest again, do not rely on a physical directory structure with the MEF format. This was the root cause of the problem, if I remember correctly. I am sorry I do not remember all the details. It has been 9 months since my involvement in the project relying on MEF ended, most probably for good.

Situation on Linux is a little bit more complicated because many more file systems are supported by that OS.

**Comment by dancrepeau** [comment]
The following code is what I propose.  _MAX_PATH in Windows is 260, and it differs in unix.  I just tested PATH_MAX on Mac OS and got 1024, and I think it's higher on linux (4096).  I am multiplying by 4 because that is the greatest byte size of a unicode character (in both UTF-8 and UTF-16).  I understand that it is possible to prefix a path to use a longer string - my opinion is that this is beyond what most users are going to need.  I also want to emphasize that the MEF format does not rely on physical directory structure (in the sense that the complete path is not stored in any of the MEF files).  The meflib code is a set of helper routines, and can be modified by any user to suit specific project needs.

#ifdef _WIN32
#define MEF_FULL_FILE_NAME_BYTES		_MAX_PATH * 4
#else
#define MEF_FULL_FILE_NAME_BYTES		PATH_MAX * 4
#endif

**Comment by em7** [comment]
"my opinion is that this is beyond what most users are going to need" This is quite a bold statement and is usually considered to be a very bad engineering practice. :-)

If I remember correctly, the MEF "file" was actually a directory with an extension ".mefd" and contained a set of physical directories and files, multiple per signal stored. This is, from software engineering point of view, reliance on a physical file/directory structure.

I'm afraid of that what you suggest would not help at all.

**Comment by dancrepeau** [comment]
I am sincerely trying to help here.  Your original comment was that any path longer than 256 characters failed to read.  I could certainly set up a testbed and try to reproduce the problem.  Perhaps if you could give an idea of how many characters you'd like to support?  We could hard-code 32768 in, although you don't like hard-coding.  When I proposed using _MAX_PATH, you don't like that either, since the number is too small on Windows.  I would suspect that in addition to raising the buffer size, we'd also need to make sure proper versions of fopen() and other functions would need to be used.  We are not opposed to making these changes, but might need someone with the proper expertise to lead the effort.  It's still not clear to me whether the issue is just really long path names, or whether it's a unicode-related issue.

**Comment by em7** [comment]
I really appreciate your effort. Unfortunately my memory is a bit rusty, it has been 9 months since I was involved in using MEF and I no longer have an access to the source codes of the software in question nor the data. During this conversation I was able to get in touch with nelagri and from our "collective memory" it looks like the proposed change might not be enough.

Speaking of long file names (since Windows uses a bit of an ... interesting? :-) ... way of handling them), you can definitely use this function from Windws API https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew . Or use an already existing multiplatform C++ library rather than plain C and ANSI C APIs which are ... well, just a human-readable Assembly code.

I wrote a lot of C code and even some C++ during my professional software engineering career. However, since I am first and foremost a Java programmer, I'm definitely not the right person to be expected to have a proper level of expertise to engineer the MEF library.

**Comment by dancrepeau** [comment]
Agreed that Windows' way of doing file paths is "interesting".  Well, I just went ahead and updated meflib.h with the change I proposed earlier in this thread.  I doubt it will make much difference (except possibly on linux, which seems to have a much higher value for PATH_MAX).  But this shouldn't hurt anything either.

I think the next step if we want to pursue this further is to set up a testbed with very long paths, and figure out which routines are crashing and need updating.  Looks like you started to do this when you introduced the thread, with line 9220 of meflib.c being called out.

Thank you for calling this issue to our attention - it is an area that needs some work.
